Feature/#44 Setup Room

khaledeid1k opened the pull request:
feat: Setup Room

kareem-01 committed:
feat: add shared UI components

kareem-01 committed:
Merge branch 'develop' into feature/shared-components

kareem-01 committed:
feat: update TaskUiState to use String for taskId and add taskDate

kareem-01 committed:
feat: add calendar icon and display task date in TaskCard

khaledeid1k committed:
Add predefined image and title of categories

khaledeid1k committed:
feat: Implement local database with Room

This commit introduces the local database setup using Room persistence library.

Key changes:
- Added `TudeeDaoCategory` interface defining Data Access Object (DAO) methods for category operations (create, delete, edit, get all, get by ID).
- Created `TudeeDataBase` abstract class, which extends `RoomDatabase`, to define the database structure and provide access to the DAO.
- Implemented a singleton pattern for `TudeeDataBase` to ensure only one instance of the database is created.
- Added `Constants.kt` file to store the database name (`TudeeDatabase`).

khaledeid1k committed:
Refactor: Update KSP version and align Room KTX version reference

The KSP (Kotlin Symbol Processing) version has been updated from `2.1.21-2.0.2` to `2.0.21-1.0.27`.
Additionally, the version reference for `androidx-room-ktx` has been changed from `roomKtx` to `roomVersion` for consistency, as both pointed to the same version `2.7.1`.

khaledeid1k committed:
feat: implement LocalDataSource and CategoryEntity

This commit introduces the `LocalDataSource` interface and its implementation `LocalDataSourceIm`, which interacts with `TudeeDaoCategory` for category data operations.

It also defines the `CategoryEntity` data class, representing a category in the local database with properties for ID, title, image, and task count.

The `CategoriesServiceImpl` has been updated to accept a `LocalDataSource` instance.

khaledeid1k committed:
Merge branch 'feature/#43-categories-ui' into test/#44_setup_room

# Conflicts:
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/AppTextField.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/CancelableActionLayout.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/CategoryCard.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskCard.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/dayCard.kt
#	app/src/main/res/values/strings.xml

khaledeid1k committed:
feat: Integrate local data source for categories and update dependencies

This commit introduces the integration of a local data source for managing categories, along with necessary dependency updates and mapper implementations.

Key changes:
- Added `coil-compose` and `koin-androidx-compose` dependencies to `gradle/libs.versions.toml` and `app/build.gradle.kts`.
- Modified `Category` domain entity: `id` type changed from `UUID` to `Long`, and `image` field renamed to `icon`.
- Implemented `dataBaseModule` for Koin dependency injection, providing `TudeeDaoCategory` and `LocalDataSourceIm`.
- Updated `serviceModule` to include `CategoriesServiceImpl` binding.
- Added `appModule` to include `dataBaseModule`.
- Created `CategoryMapper.kt` in `presentation.categories.mapper` for mapping between domain `Category` and `CategoriesScreenStatus.Category`.
- Created `CategoryMapper.kt` in `data.localSource.mapper.category` for mapping between domain `Category` and `CategoryEntity`.
- Introduced `CategoriesScreenStatus` data class to represent category data on the UI.
- Updated `LocalDataSource` and `LocalDataSourceIm` to reflect changes in `CategoryEntity` (specifically, `deleteCategory` now takes `CategoryEntity` instead of an ID).
- Updated `CategoriesService` interface to align with `Category` entity changes (ID type and `deleteCategory` parameter).
- Removed an unused "cancel" string resource.

khaledeid1k committed:
Refactor: Update TudeeDaoCategory methods to suspend functions

The methods within the `TudeeDaoCategory` interface (`createCategory`, `deleteCategory`, `editCategory`, `getAllCategories`, and `getCategoryById`) have been modified to be `suspend` functions. This change enables them to be called from coroutines, allowing for asynchronous database operations without blocking the main thread.

Additionally, the `coil-compose` library has been added to the project dependencies with version `2.7.0`.

khaledeid1k committed:
Refactor: Update AppTheme import paths

The import path for `AppTheme` has been changed from `com.washingtondcsquad.tudee.presentation.design.theme.AppTheme` to `com.washingtondcsquad.tudee.presentation.design.AppTheme` in several composable files. This change standardizes the import path for the application's theme.

khaledeid1k committed:
Refactor: Update AppTheme import path

The import path for `AppTheme` in `CategoryCard.kt` has been changed from `com.washingtondcsquad.tudee.presentation.design.theme.AppTheme` to `com.washingtondcsquad.tudee.presentation.design.AppTheme`.

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/TudeeDataBase.kt:
```diff
@@ -0,0 +1,31 @@
+import com.washingtondcsquad.tudee.data.localSource.entities.CategoryEntity
+import com.washingtondcsquad.tudee.data.utils.Constants
+
+@Database(entities = [CategoryEntity::class], version = 1, exportSchema = false)
```
we need make a table for tasks

khaledeid1k replied:
you can add like @Database(entities = [CategoryEntity::class,Tasks::class, version = 1, exportSchema = false)

everyone is responsible for add table of his task at room as we talked

MostafaMohamed2002 replied:
now that's clear , Thanks

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/TudeeDataBase.kt:
we don't need all of this logic because we can provide our instance from Room database and our dao's using **Koin**

khaledeid1k replied:
i don't understand what do you mean

MostafaMohamed2002 replied:
you can check that 
https://stackoverflow.com/a/69835118

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/data/services/LocalDataSource.kt:
```diff
@@ -0,0 +1,11 @@
+
+import com.washingtondcsquad.tudee.data.localSource.entities.CategoryEntity
+
+interface LocalDataSource {
```
local data source of what ?

khaledeid1k replied:
that of meaning of abstraction for if I change room library with new library not affect to our code

omer1998 replied:
yes khalid but at least specify that this for category 
like this : CategoryLocalDataSource
in order to create other one for task

MostafaMohamed2002 replied:
sorry , but i mean the name is not clear

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/data/utils/Constants.kt:
```diff
@@ -0,0 +1,5 @@
+package com.washingtondcsquad.tudee.data.utils
+
+object Constants {
```
using Constant as name is more genral you can use more specific name like Database contans

khaledeid1k replied:
for if need to add another constant not related to data base

MostafaMohamed2002 commented:
please resolve the merge confilcts

omer1998 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskCard.kt:
```diff
@@ -77,7 +79,11 @@ fun TaskCard(
                 }
 
                 // to be changed when project structure is merged
-                TaskPriorityCard(fromStringToPriority(taskUiState.taskPriority))
```
yes i understand, but it is always appropriate to work on the latest develop branch
just pull the latest develop and merge it with your new code and everything will be fine

omer1998 reviewed (commented):
you've done a great work khalid thanx for your efforts
also it would be better to separate this pr into two part one for room setup and the other for category it would be more organaized and easy to be reviewed

Mostafa-alsaygh commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/LocalDataSourceIm.kt:
```diff
@@ -0,0 +1,33 @@
+import com.washingtondcsquad.tudee.data.localSource.entities.CategoryEntity
+import com.washingtondcsquad.tudee.data.services.LocalDataSource
+
+class LocalDataSourceIm (
```
there is problem with the name

Mostafa-alsaygh replied:
LocalDataSourceImpl

Mostafa-alsaygh commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/TudeeDao.kt:
```diff
@@ -0,0 +1,30 @@
+import com.washingtondcsquad.tudee.data.utils.Constants
+
+@Dao
+interface TudeeDaoCategory {
```
no need on each name add tudee

Mostafa-alsaygh commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/entities/CategoryEntity.kt:
```diff
@@ -0,0 +1,16 @@
+@Entity(tableName = Constants.ROOM_DATABASE_NAME)
+data class CategoryEntity(
+    @PrimaryKey(autoGenerate = true) val id: Long,
+    @ColumnInfo(name = "Category Title") val title: String,
```
no need for each parameter a  @ColumnInfo
why do u use it

Mostafa-alsaygh commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/mapper/category/CategoryMapper.kt:
```diff
@@ -0,0 +1,24 @@
+    )
+}
+
+fun CategoryEntity.toCategory(): Category {
```
what is a different between a  toCategory and toCategoryEntity
i thing u can name it better 
like adding local word

HendSayed25 reviewed (approved):
great work khaleid !

khaledeid1k committed:
Merge branch 'develop' into feature/#44_setup_room

# Conflicts:
#	app/src/main/res/values/strings.xml

khaledeid1k committed:
Refactor: Rename LocalDataSourceIm and TudeeDaoCategory

This commit renames `LocalDataSourceIm` to `LocalDataSourceImpl` for better adherence to naming conventions.

Additionally, `TudeeDaoCategory` has been renamed to `DaoCategory` for brevity and clarity. All corresponding references in `TudeeDataBase.kt`, `LocalDataSourceImpl.kt`, and `dataBaseModule.kt` have been updated.

khaledeid1k committed:
Refactor: Rename LocalDataSource to CategoryLocalDataSource

This commit renames `LocalDataSource` and its implementation `LocalDataSourceImpl` to `CategoryLocalDataSource` and `CategoryLocalDataSourceImpl` respectively.

This change provides more specific naming for the data source responsible for category-related operations.

The dependencies in `CategoriesServiceImpl` and Koin modules have been updated accordingly.

khaledeid1k committed:
Refactor: Rename Constants to DataBaseConstants and update mapper function names

This commit renames the `Constants` object to `DataBaseConstants` to better reflect its purpose.

Additionally, the category mapper functions have been renamed for clarity:
- `toCategoryEntity()` is now `toEntity()`
- `toCategory()` is now `toDomain()`

A minor comment has also been removed from `TaskCard.kt`.

khaledeid1k committed:
Refactor: Simplify `dropShadow` Modifier Extension

This commit refactors the `dropShadow` Modifier extension for improved readability and conciseness.

Key changes:
- Removed unnecessary `apply` block for `Paint` initialization.
- Simplified the `BlurMaskFilter` application.
- Removed an unnecessary calculation for `spreadPx` when translating the canvas.
